File: 02_BFS/Gyeongwook/atoi.py
class Solution:
    # @return an integer
    def atoi(self, str):
        strLen = len(str)
        
        if strLen == 0 :
            return 0
            
        sign = 1
        startPos = 0
        result = 0
        
        while str[startPos] == ' ' :
            startPos = startPos + 1
        
        if str[startPos] == '-' or str[startPos] == '+' :
            if str[startPos] == '-' :
                sign =  sign * -1
                
            startPos = startPos + 1
            
        for idx in range(startPos, strLen) :
            if str[idx] < '0' or str[idx] > '9' :
                result = sign * result
                if result > 2147483647 :
                    return 2147483647
                elif result < -2147483648 :
                    return -2147483648
                return result
            
            result = result * 10
            result = result + int(str[idx])
            
        
        result = sign * result
        if result > 2147483647 :
            return 2147483647
        elif result < -2147483648 :
            return -2147483648
        return result

File: 02_BFS/Seonghun/Atoi.java
public class Solution {

	// test
    
    private enum State{
        BEFORE,
        SIGN,
        INT
    }
    
    public int atoi(String str) {
        
        State state = State.BEFORE;
        
        char ch;
        
        int size = str.length();
        int idx = 0;
        int sign = 1;
        
        StringBuilder sb = new StringBuilder();
        
        while(idx < size){
            ch = str.charAt(idx);
            switch(state){
                case BEFORE:
                    if('0' <= ch && ch <= '9'){
                        sb.append(ch);
                        state = State.INT;
                    }
                    else if(ch == '+'){
                        sb.append(ch);
                        state = State.SIGN;
                    }
                    else if(ch == '-'){
                        sb.append(ch);
                        state = State.SIGN;
                        sign = -1;
                    }
                    else if(ch != ' '){
                        return 0;
                    }
                    break;
                case SIGN:
                    if('0' <= ch && ch <= '9'){
                        sb.append(ch);
                        state = State.INT;
                    }
                    else {
                        return 0;
                    }
                    break;
                case INT:
                    if('0' <= ch && ch <= '9'){
                        sb.append(ch);
                    }
                    else {
                        idx = size; // for escaping loop
                    }
                    break;
                default:
                    break;
            }
            idx++;
        }
        
        try{
            if(sb.length() > 0)   return Integer.parseInt(sb.toString());
            else                  return 0;    
        }
        catch(NumberFormatException e){ 
            return (sign == 1)? Integer.MAX_VALUE:Integer.MIN_VALUE;
        }
        
    }

}
